presto-main/src/main/java/io/prestosql/server/protocol/QueryResultRows.java

```diff
@@ -184,6 +179,48 @@ else if (type instanceof TimestampWithTimeZoneType) {
         return Optional.of(unmodifiableList(row));
     }
 
+    private Object roundParametricDateTime(Object value, Type type)
+    {
+        if (supportsParametricDateTime || value == null) {
+            return value;
+        }
+
+        if (type instanceof TimestampType) {
+            return ((SqlTimestamp) value).roundTo(3);
+        }
+
+        if (type instanceof TimestampWithTimeZoneType) {
+            return ((SqlTimestampWithTimeZone) value).roundTo(3);
+        }
+
+        if (type instanceof ArrayType) {
+            List<Object> values = (List<Object>) value;
+
+            ImmutableList.Builder<Object> result = ImmutableList.builder();
+            Type elementType = ((ArrayType) type).getElementType();
+
+            for (int i = 0; i < values.size(); i++) {
+                result.add(roundParametricDateTime(values.get(i), elementType));
+            }
+
+            return result.build();
+        }
+
+        if (type instanceof MapType) {
+            Map<Object, Object> values = (Map<Object, Object>) value;
+
+            ImmutableMap.Builder<Object, Object> result = ImmutableMap.builder();
+            Type keyType = ((MapType) type).getKeyType();
+            Type valueType = ((MapType) type).getValueType();
+
+            values.entrySet().forEach(entry -> result.put(roundParametricDateTime(entry.getKey(), keyType), roundParametricDateTime(entry.getValue(), valueType)));
+
+            return result.build();
+        }
+
+        return value;
```

Comment:
Does ROW type also need to be handled here?
i guess not, but worth adding a comment why

Reply:
It needs as @martint pointed me. I'm working on updated PR :)